clickhouse_connect/driver/transform.py

```diff
@@ -1,38 +1,116 @@
 
 
-class DataTransform(ABC):
-
-    def parse_response(self, source: Sequence, context: QueryContext = _EMPTY_QUERY_CONTEXT) -> DataResult:
-        """
-        Decodes the ClickHouse byte buffer response into rows of native Python data
-        :param source: A byte buffer or similar source
-        :param context: The QueryContext to use in processing the response
-        :return: DataResult -- data matrix, column names, column types
-        """
+class NativeTransform:
+    # pylint: disable=too-many-locals
+    @staticmethod
+    def parse_response(source: ByteSource, context: QueryContext = _EMPTY_QUERY_CONTEXT) -> QueryResult:
         with context:
-            return self._transform_response(source, context)
-
-    def build_insert(self, context: InsertContext):
-        """
-        Encodes a dataset of Python sequences into a ClickHouse format using the specified insert context
-        :param context InsertContext parameter object
-        :return: generator of bytes like objects containing the dataset in the appropriate format
-        """
+            names = []
+            col_types = []
+            use_none = context.use_none
+            block_num = 0
+
+            def get_block():
+                nonlocal block_num
+                try:
+                    num_cols = source.read_leb128()
+                    num_rows = source.read_leb128()
+                except (StopIteration, IndexError):
+                    source.close()
+                    return None
+                result_block = []
+                for col_num in range(num_cols):
+                    name = source.read_leb128_str()
+                    if block_num == 0:
+                        names.append(name)
+                    type_name = source.read_leb128_str()
+                    if block_num == 0:
+                        col_type = registry.get_from_name(type_name)
+                        col_types.append(col_type)
+                    else:
+                        col_type = col_types[col_num]
+                    context.start_column(name, col_type)
+                    column = col_type.read_native_column(source, num_rows, use_none=use_none)
+                    result_block.append(column)
+                block_num += 1
+                return result_block
+
+            first_block = get_block()
+            if first_block is None:
+                return QueryResult([])
+
+            def gen():
+                yield first_block
+                while True:
+                    next_block = get_block()
+                    if next_block is None:
+                        return
+                    yield next_block
+
+        return QueryResult(None, gen(), tuple(names), tuple(col_types), context.column_oriented, source)
+
+    @staticmethod
+    def build_insert(context: InsertContext):
+        if context.compression == 'gzip':
+            compressor = GzipCompressor()
+        else:
+            compressor = NullCompressor()
```

Comment:
do we need to create new instances here every time? If we have a single thread anyway, probably it's safe to flush Zlib from a static object

Reply:
`context.compression` could be pass with an actual compressor instance

Reply:
Going with Serge here, we only need one instance. :)